pennylane/transforms/decompose.py

```diff
@@ -60,7 +61,53 @@ def _operator_decomposition_gen(
             )
 
 
-@transform
+def _decompose_plxpr_transform(
+    self, primitive, tracers, params, targs, tkwargs, state=None
+):  # pylint: disable=unused-argument, too-many-arguments, too-many-positional-arguments
+    """Implementation for ``decompose`` with program capture"""
+    # pylint: disable=import-outside-toplevel
+    from pennylane.capture import TransformTracer
+
+    gate_set = tkwargs.pop("gate_set", set(qml.ops.__all__))
+
+    if isinstance(gate_set, (str, type)):
+        gate_set = set([gate_set])
+
+    if isinstance(gate_set, Iterable):
+        gate_types = tuple(gate for gate in gate_set if isinstance(gate, type))
+        gate_names = set(gate for gate in gate_set if isinstance(gate, str))
+        gate_set = lambda op: (op.name in gate_names) or isinstance(op, gate_types)
+
+    max_expansion = tkwargs.pop("max_expansion", None)
```

Comment:
Should we have a way to do this kind of "transform setup" once at capture time?

Reply:
That's a good idea. It might be worth adding a `setup_transform(targs, tkwargs)` function that can be called  to do this setup in `TransformInterpreter.setup`.